### PR TITLE
Allow reply and hello in callbacks to be taken by the user

### PR DESCRIPTION
**Pull request by cguimaraes**
It includes moving the z_hello_array_t into a z_hello_list_t.

### CODE REVIEW

File: src/api/api.c [hunk 1]
```diff
@@ -238,7 +242,7 @@ _MUTABLE_OWNED_FUNCTIONS_DEFINITION(z_bytes_t, z_owned_bytes_t, bytes, _z_bytes_
 _MUTABLE_OWNED_FUNCTIONS_DEFINITION(z_string_t, z_owned_string_t, string, _z_string_free, _z_owner_noop_copy)
 _IMMUTABLE_OWNED_FUNCTIONS_DEFINITION(z_keyexpr_t, z_owned_keyexpr_t, keyexpr, _z_keyexpr_free, _z_keyexpr_copy)
 _MUTABLE_OWNED_FUNCTIONS_DEFINITION(z_config_t, z_owned_config_t, config, _z_config_free, _z_owner_noop_copy)
-_MUTABLE_OWNED_FUNCTIONS_DEFINITION(z_scouting_config_t, z_owned_scouting_config_t, scouting_config, _z_owner_noop_free, _z_owner_noop_copy)
+_MUTABLE_OWNED_FUNCTIONS_DEFINITION(z_scouting_config_t, z_owned_scouting_config_t, scouting_config, _z_config_free, _z_owner_noop_copy)
```
Comment (Mallets) [inline src/api/api.c:245]: Shouldn't be `_z_scouting_config_free`?

**Review by Mallets — CHANGES_REQUESTED** [review]

File: src/api/api.c (hunk 1 above)
Comment (cguimaraes) [reply src/api/api.c]: Done.

File: src/session/query.c [hunk 2]
```diff
@@ -213,19 +219,12 @@ int _z_trigger_query_reply_final(_z_session_t *zn, const _z_reply_context_t *rep
...
-        _z_pending_reply_list_t *pen_rps = pen_qry->_pending_replies;
-        _z_pending_reply_t *pen_rep = NULL;
-        while (pen_rps != NULL) {
-            pen_rep = _z_pending_reply_list_head(pen_rps);
+        while (pen_qry->_pending_replies != NULL) {
+            _z_pending_reply_t *pen_rep = _z_pending_reply_list_head(pen_qry->_pending_replies);
 
-            // Check if this is the same resource key
             // Trigger the query handler
-            if (_z_keyexpr_intersect(pen_qry->_key._suffix, strlen(pen_qry->_key._suffix),
```
Comment (Mallets) [inline src/session/query.c:223]: I believe this check shouldn't be removed to guarantee that all the replies are for the query being issued.

**Review by Mallets — CHANGES_REQUESTED** [review]

File: examples/c11/z_scout.c [hunk 3]
```diff
@@ -64,8 +64,8 @@ void fprinthello(FILE *stream, const z_hello_t *hello) {
     fprintf(stream, " }");
 }
 
-void callback(z_owned_hello_t hello, void *context) {
-    fprinthello(stdout, hello._value);
+void callback(z_owned_hello_t *hello, void *context) {
+    fprinthello(stdout, hello->_value);
```
Comment (p-avital) [inline examples/c11/z_scout.c:68]: not necessary for this PR, but we need a loan function here (zenoh-c currently doesn't have the equivalent loan-type)

File: examples/c99/z_scout.c [hunk 4]
```diff
@@ -62,8 +62,8 @@ void fprinthello(FILE *stream, const z_hello_t *hello) {
     fprintf(stream, " }");
 }
 
-void callback(z_owned_hello_t hello, void *context) {
-    fprinthello(stdout, hello._value);
+void callback(z_owned_hello_t *hello, void *context) {
+    fprinthello(stdout, hello->_value);
```
Comment (p-avital) [inline examples/c99/z_scout.c:66]: same as c11 scouting example

File: src/session/query.c (hunk 2 above)
Comment (p-avital) [reply src/session/query.c:223]: As discussed internally, agreed

Comment (cguimaraes) [reply src/session/query.c:223]: I agree. We must ensure that the user do not get a reply outside the requested set of keyexpr.

**Review by p-avital — APPROVED** [review]

File: examples/c99/z_scout.c (hunk 4 above)
Comment (cguimaraes) [reply examples/c99/z_scout.c]: In Pico we have it, so I am just using the z_loan now.

File: examples/c11/z_scout.c (hunk 3 above)
Comment (cguimaraes) [reply examples/c11/z_scout.c]: In Pico we have it, so I am just using the z_loan now.

**Review by Mallets — APPROVED** [review]

**Pull request merged**